### PR TITLE
feat: setup and configuration for redis server

### PR DESCRIPTION
**Pull request by PeterStrob**
Signed-off-by: Peter Strobel <peter@indicio.tech>

### CODE REVIEW

File: int/docker-compose.yml [hunk 1]
```diff
@@ -51,7 +71,7 @@ services:
       dockerfile: ./Dockerfile.test.runner
     environment:
       - WAIT_BEFORE_HOSTS=3
-      - WAIT_HOSTS=acapy_pickup:3000,echo:3002
+      - WAIT_HOSTS=acapy_pickup:3000,echo:3002,redis:6379
```
Comment (burdettadam) [inline int/docker-compose.yml:74]: is adding redis to the wait_hosts needed with the health check?

File: docker/docker-compose.yml [hunk 2]
```diff
@@ -28,3 +28,27 @@ services:
...
+  redis:
+    image: "redis:alpine"
+
+    ports:
+     - "6379:6379"
+
+    volumes:
+      - ./redis-data:/var/lib/redis
+      - ./redis.conf:/etc/redis/redis.conf
+
```
Comment (burdettadam) [inline docker/docker-compose.yml:40]: does Redis need a health check? should the agent depend on Redis?

File: redis.conf [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+protected-mode yes
+port 6379
+
+timeout 0
+
+tcp-keepalive 300
+
+################################### GENERAL ###################################
+
+daemonize-no
```
Comment (burdettadam) [inline redis.conf:14]: should it be `daemonize no` instead of `daemonize-no`?

**Review by burdettadam — CHANGES_REQUESTED** [review]
I think docker-compose could be improved a little. I don't see any other issues, so if tests pass I approve.

File: docker/docker-compose.yml (hunk 2 above)
Comment (PeterStrob) [reply docker/docker-compose.yml:40]: The health check is something Daniel recommended be added in, I think it's not a bad idea to have one

File: redis.conf (hunk 3 above)
Comment (PeterStrob) [reply redis.conf:14]: You are right!

File: int/docker-compose.yml (hunk 1 above)
Comment (PeterStrob) [reply int/docker-compose.yml:74]: No it is not, I think this was added in to check that the server was connecting properly

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -18,6 +18,8 @@ jobs:
 
     steps:
     - uses: actions/checkout@v2
+    - name: Redis Server in GitHub Actions
+      uses: supercharge/redis-github-action@1.4.0
```
Comment (dbluhm) [inline .github/workflows/tests.yml:22]: This should not be necessary. Integration test docker compose should spin up the redis server itself. Unit tests (tests run outside of the int test docker containers) should not rely on redis at all (mocked interfaces or similar should be used).

File: acapy_plugin_pickup/protocol/delivery.py [hunk 5]
```diff
@@ -153,76 +153,104 @@ async def remove_message_for_key(self, key: str, msg: OutboundMessage):
         """Remove specified message from queue for key."""
 
 
-class InMemoryQueue(UndeliveredInterface):
-    """In-memory undelivered message queue."""
```
Comment (dbluhm) [inline acapy_plugin_pickup/protocol/delivery.py:157]: The `InMemoryQueue` should not have been deleted in favor of the `PersistedQueue`. Both can coexist and the appropriate implementation for the use case should be selected by configuration.

File: acapy_plugin_pickup/protocol/delivery.py [hunk 6]
```diff
@@ -153,76 +153,104 @@ async def remove_message_for_key(self, key: str, msg: OutboundMessage):
...
+class PersistedQueue(UndeliveredInterface):
```
Comment (dbluhm) [inline acapy_plugin_pickup/protocol/delivery.py:156]: `PersistedQueue` should be something like `RedisUndeliveredQueue`. `PersistedQueue` implies that it's a generic/ABC for all Queues providing persistence.

File: docker/docker-compose.yml [hunk 7]
```diff
@@ -28,3 +28,27 @@ services:
...
+      node_net:
+        ipv4_address: 172.28.1.4
+
+# networking for the Redis container
+networks:
+  node_net:
+    ipam:
+      driver: default
+      config:
+        - subnet: 172.28.0.0/16
```
Comment (dbluhm) [inline docker/docker-compose.yml:54]: Separate networking not needed.

File: docker/docker-compose.yml (hunk 2 above)
Comment (dbluhm) [reply docker/docker-compose.yml:40]: Redis should have a healthcheck and then the agent should have a depends on condition of `service_healthy` for the service.

File: int/docker-compose.yml (hunk 1 above)
Comment (dbluhm) [reply int/docker-compose.yml:74]: We need to stick with either health checks or wait scripts. We've got healthchecks so I think we should commit to that, making these `WAIT_` vars unneeded. Dockerfile for the juggernaut would also need to be updated to drop the addtion of the wait script.

File: int/docker-compose.yml [hunk 8]
```diff
@@ -62,3 +82,5 @@ services:
     depends_on:
       - acapy_pickup
       - echo
+      - redis
+
```
Comment (dbluhm) [inline int/docker-compose.yml:86]: These need service healthy conditions.

File: tests/test_persistedqueue.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+
+@pytest.mark.asyncio
+async def test_persistedqueue(msg):
+    """
+    PersistedQueue Test.
+
+    Unit test for the PersistedQueue class in
+    """
+    PQ = PersistedQueue(redis=await aioredis.from_url("redis://localhost/1"))
```
Comment (dbluhm) [inline tests/test_persistedqueue.py:37]: This is one of our problems. This test should be in the integration tests, not in the unit tests. Tests that depend on external services are by definition integration tests or "higher" in the testing hierarchy (unit < int < system < ....)

File: tests/test_persistedqueue.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
+import aioredis
+
+from acapy_plugin_pickup.protocol.delivery import PersistedQueue
+from aries_cloudagent.transport.outbound.message import OutboundMessage
+import pytest
+
+
+@pytest.fixture
+def target():
+    yield {
```
Comment (dbluhm) [inline tests/test_persistedqueue.py:10]: The object yielded needs to be a `ConnectionTarget`.

File: tests/test_persistedqueue.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    yield OutboundMessage(
+        connection_id="conn_id",
+        target=target,
+        reply_from_verkey="reply_from_verkey",
+        payload="payload",
+    )
+
+
+@pytest.mark.asyncio
+async def test_persistedqueue(msg):
```
Comment (dbluhm) [inline tests/test_persistedqueue.py:31]: This test needs some work to better reflect how the queue will actually work.

**Review by dbluhm — CHANGES_REQUESTED** [review]
Some more items to work on. We also need to reconcile the different branches that now exist for redis related changes.

We need integration tests that actually use the persisted queue for storing undelivered messages. This means that the delivery req handlers need to use a persisted queue instance and we need event handlers for capturing the undelivered messages and pushing them to the queue.

**Comment by dbluhm** [comment]
Just left a bunch of review comments but maybe we need break these changes into smaller pieces. In general, we need to push for smaller, incremental changes.

**Pull request closed without merging**